### DISCUSSION TITLE
Example of using SetWinEventHook?

### DISCUSSION THREAD

**Discussion by skint007** [microsoft/CsWin32#1162] [Q&A]
I can't seem to get `SetWinEventHook` to work, it never triggers the delegate. After I create the hook, the handle that is returned says it's "`IsValid`".

Heres how I setup the hook
```csharp
private const int WINEVENT_OUTOFCONTEXT = 0x0000;
private const int EVENT_SYSTEM_FOREGROUND = 0x0003;
_hForegroundHook = PInvoke.SetWinEventHook(EVENT_SYSTEM_FOREGROUND, EVENT_SYSTEM_FOREGROUND, null, WinEventProc, 0, 0, WINEVENT_OUTOFCONTEXT);
```

This is the method I'm expecting to be triggered
```csharp
private void WinEventProc(HWINEVENTHOOK hWinEventHook, uint eventType, HWND hwnd, int idObject, int idChild, uint dwEventThread, uint dwmsEventTime)
{
    // Some code
}
```

I'm sure I'm missing something but I'm not sure what.

Any hints are appreciated :)

**Comment by AArnott** [comment]
`WinEventProc` as you've defined it is an _instance_ method. But native function pointers don't include a `this` pointer like C# Delegates do. It'll have to be static. 
Then there's the fact that you'll need to pin the delegate so that the GC doesn't move it in memory while native code has a pointer to it.

**Reply by skint007** [reply]
Thank you, I must have missed those points when looking up examples, I will make those changes 😊
